diff --git a/js/panels_ajax_tab.js b/js/panels_ajax_tab.js
index 015fcb7..3d5c8b7 100644
--- a/js/panels_ajax_tab.js
+++ b/js/panels_ajax_tab.js
@@ -41,7 +41,10 @@
           else {
             currentTab = tabs.filter('*[data-panel-name="' + preloaded + '"]');
             // Prime the cache from the preloaded content
-            currentTab.data('panels-ajax-tab-cache', $('#panels-ajax-tab-container-' + target_id).html());
+            var cache = {
+                markup: $('#panels-ajax-tab-container-' + target_id).html(),
+            };
+            currentTab.data('panels-ajax-tab-cache', cache);
           }
           
           currentTab.addClass('panels-ajax-tabs-first-loaded');
@@ -90,9 +93,15 @@
         
         // If we have it cached we don't need to do AJAX
         if (typeof $tab.data('panels-ajax-tab-cache') !== "undefined") {
-          $('#panels-ajax-tab-container-' + target_id).html($tab.data('panels-ajax-tab-cache'));
+          $('#panels-ajax-tab-container-' + target_id).html($tab.data('panels-ajax-tab-cache').markup);
           Drupal.attachBehaviors($('#panels-ajax-tab-container-' + target_id)[0]);
           
+          // Merge the js settings
+          if ($tab.data('panels-ajax-tab-cache').js_settings != undefined) {
+            var settings = $.extend({}, Drupal.settings, $tab.data('panels-ajax-tab-cache').js_settings);
+            Drupal.settings = settings;
+          }
+          
           loadedEvent.data.cached = true;
           
           $(container).data('loading', false);
@@ -119,12 +128,19 @@
               $(container).data('loading', false);
             }
           }).done(function(data) {
-            $('#panels-ajax-tab-container-' + target_id).html(data);
+            $('#panels-ajax-tab-container-' + target_id).html(data['markup']);
             Drupal.attachBehaviors($('#panels-ajax-tab-container-' + target_id)[0]);
             $(container).data('loading', false);
             
+            // Merge the js settings
+            if (data.js_settings != undefined) {
+              var settings = $.extend({}, Drupal.settings, data.js_settings);
+              Drupal.settings = settings;
+
+            }
+            
             // Cache the contents
-            $tab.data('panels-ajax-tab-cache', $('#panels-ajax-tab-container-' + target_id).html());
+            $tab.data('panels-ajax-tab-cache', data);
 
             // Trigger optional callback
             if (callback) {
diff --git a/panels_ajax_tab.module b/panels_ajax_tab.module
index 9eef45b..7aef92b 100644
--- a/panels_ajax_tab.module
+++ b/panels_ajax_tab.module
@@ -112,8 +112,22 @@ function panels_ajax_tab_ajax($panel_name, $context_string, $url_enabled) {
   drupal_static_reset('drupal_add_js');
   drupal_static_reset('drupal_add_css');
 
-  print theme('panels_ajax_tab_ajax', array('minipanel' => $mini, 'url_enabled' => $url_enabled));
 
+
+  $output = theme('panels_ajax_tab_ajax', array('minipanel' => $mini, 'url_enabled' => $url_enabled));
+
+  // Get the settings so we can merge them in the callback
+  $javascript = drupal_add_js();
+
+  $response = array();
+  $response['markup'] = $output;
+
+  // Add settings if there are some
+  if (!empty($javascript['settings'])) {
+    $response['js_settings'] = $javascript['settings'];
+  }
+
+  drupal_json_output($response);
   exit;
 }
 
@@ -211,6 +225,10 @@ function theme_panels_ajax_tab_ajax($vars) {
   // Get the javascript and filter out the default misc/drupal.js - it's already been added by the calling page
   $javascript = drupal_add_js();
   unset($javascript['misc/drupal.js']);
+
+  // Settings are handled in the ajax page callback
+  unset($javascript['settings']);
+
   $output .= drupal_get_js('header', $javascript);
 
   // Add CSS
